File: gems/event_stream/lib/event_stream.rb
# frozen_string_literal: true

#
# Copyright (C) 2014 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

require "active_support"
require "active_record"
require "bookmarked_collection"
require "canvas_cassandra"
require "inst_statsd"

module EventStream
  require "event_stream/attr_config"
  require "event_stream/backend"
  require "event_stream/record"
  require "event_stream/failure"
  require "event_stream/logger"
  require "event_stream/stream"
  require "event_stream/index"

  def self.current_shard
    @current_shard_lookup&.call
  end

  def self.current_shard_lookup=(callable)
    @current_shard_lookup = callable
  end

  def self.get_index_ids(index, rows)
    @get_index_ids_lookup ||= ->(index2, rows2) { rows2.pluck(index2.id_column) }
    @get_index_ids_lookup.call(index, rows)
  end

  def self.get_index_ids_lookup=(callable)
    @get_index_ids_lookup = callable
  end
end

File: client_apps/canvas_quizzes/apps/events/js/views/event_stream.js
/** @jsx React.DOM */
/*
 * Copyright (C) 2014 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

define(function(require) {
  var React = require('old_version_of_react_used_by_canvas_quizzes_client_apps');
  var _ = require('lodash');
  var Actions = require('../actions');
  var I18n = require('i18n!quiz_log_auditing.event_stream').default;
  var ScreenReaderContent = require('jsx!canvas_quizzes/components/screen_reader_content');
  var Event = require('jsx!./event_stream/event');
  var K = require('../constants');
  var visibleEventTypes = [
    K.EVT_SESSION_STARTED, K.EVT_QUESTION_ANSWERED, K.EVT_QUESTION_VIEWED,
    K.EVT_QUESTION_FLAGGED, K.EVT_PAGE_BLURRED, K.EVT_PAGE_FOCUSED
  ];

  var extend = _.extend;

  var EventStream = React.createClass({
    getDefaultProps: function() {
      return {
        events: [],
        submission: {},
        questions: []
      };
    },

    render: function() {
      var visibleEvents = this.getVisibleEvents(this.props.events);
      return(
        <div id="ic-EventStream">
          <h2>{I18n.t('headers.action_log', 'Action Log')}</h2>

          {visibleEvents.length === 0 &&
            <p>
              {I18n.t('notices.no_events_available',
                'There were no events logged during the quiz-taking session.'
              )}
            </p>
          }

          <ol id="ic-EventStream__ActionLog">
            {visibleEvents.map(this.renderEvent)}
          </ol>
        </div>
      );
    },

    renderEvent: function(e) {
      var props = extend({}, e, {
        startedAt: this.props.submission.startedAt,
        questions: this.props.questions,
        attempt: this.props.attempt
      });

      return Event(props);
    },

    getVisibleEvents: function(events) {
      return events.filter(function(e) {
        if(visibleEventTypes.indexOf(e.type) == -1) {
          return false
        }
        if(e.type != K.EVT_QUESTION_ANSWERED) {
          return true;
        }
        return _.any(e.data, function(i) {
          return (i.answer != null);
        });
      });
    }

  });

  return EventStream;
});